src/orange/univ/detail/LazySenseCalculator.hh

```diff
@@ -0,0 +1,163 @@
+//----------------------------------*-C++-*----------------------------------//
+// Copyright 2021-2024 UT-Battelle, LLC, and other Celeritas developers.
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file orange/univ/detail/LazySenseCalculator.hh
+//---------------------------------------------------------------------------//
+#pragma once
+
+#include "corecel/Assert.hh"
+#include "corecel/cont/Range.hh"
+#include "corecel/cont/Span.hh"
+#include "orange/OrangeTypes.hh"
+#include "orange/surf/LocalSurfaceVisitor.hh"
+#include "orange/univ/detail/Types.hh"
+
+#include "SurfaceFunctors.hh"
+#include "../VolumeView.hh"
+
+namespace celeritas
+{
+namespace detail
+{
+//---------------------------------------------------------------------------//
+/*!
+ * Calculate senses with a fixed particle position.
+ *
+ * This is an implementation detail used in initialization *and* complex
+ * intersection.
+ */
+class LazySenseCalculator
+{
+  public:
+    // Construct from persistent, current, and temporary data
+    inline CELER_FUNCTION LazySenseCalculator(LocalSurfaceVisitor const& visit,
+                                              Real3 const& pos,
+                                              Span<Sense> sense_cache,
+                                              Span<SenseModFlags> sense_flags);
+
+    // Calculate senses for a single face of the given volume, possibly on a
+    // face
+    inline CELER_FUNCTION Sense operator()(VolumeView const& vol,
+                                           FaceId face_id,
+                                           OnFace face = {});
+
+    //! Clear the cached sense values
+    CELER_FUNCTION void invalidate_cache()
```

Comment:
Would it be OK (no less efficient) to invalidate every time you instantiate the LSC?

Reply:
We already kinda do that, the ctor clears all the bit flags, this only clears the cached flag